[usage] Use attribution ID to reduce DB queries for usage report

easyCZ opened the pull request:
## Description
<!-- Describe your changes in detail -->

Updates usage controller flow to do the following:

1. Find all Workspace Instances (with time bounds) which have an `usageAttributionId` set.
2. Groups these (in-code) by Attribution ID to create the report

We can't use SQL level group-by because it can only be used with summary functions (like count). The secondary reason is that the Billing controller (which runs after the usage controller) needs to given the full list of WorkspaceInstances to exclude any WorkspaceInstances which it may have already billed for (in another invoice, for example). For this, it has to get the raw data-set and can't receive an aggregation.

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
* Part of https://github.com/gitpod-io/gitpod/issues/10922

## How to test
<!-- Provide steps to test this PR -->
Unit tests
Can be run against staging by port forwarding

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->
NONE

## Werft options:
<!--
Optional annotations to add to the werft job.

* with-preview - whether to create a preview environment for this PR
-->
- [ ] /werft with-preview

roboquat commented:
@easyCZ: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

jankeromnes commented on components/usage/pkg/db/workspace_instance.go:
```diff
@@ -79,6 +79,7 @@ func ListWorkspaceInstancesInRange(ctx context.Context, conn *gorm.DB, from, to
 		).
 		Where("creationTime < ?", TimeToISO8601(to)).
 		Where("startedTime != ?", "").
+		Where("usageAttributionId != ?", "").
```
How will we want to handle potentially un-attributed instances?

easyCZ replied:
Given the `usageAttributionId` should always be present (the premise of your PR). The check here mostly ensures we only run on "new" data which does have the attribution set

jankeromnes commented on components/usage/pkg/controller/reconciler.go:
```diff
@@ -302,20 +200,20 @@ func trimStartStopTime(instances []db.WorkspaceInstance, maximumStart, minimumSt
 	WorkspaceSeconds int64  `json:"workspace_seconds"`
 }
+
+func groupInstancesByAttributionID(instances []db.WorkspaceInstance) map[db.AttributionID][]db.WorkspaceInstance {
+	result := map[db.AttributionID][]db.WorkspaceInstance{}
+	for _, instance := range instances {
+		if _, ok := result[instance.UsageAttributionID]; !ok {
+			result[instance.UsageAttributionID] = []db.WorkspaceInstance{}
+		}
+
+		result[instance.UsageAttributionID] = append(result[instance.UsageAttributionID], instance)
+	}
+
+	return result
+}
```
Are you sure it makes sense performance-wise to re-implement a SQL "group by" in Go?

I wonder if maybe this should be part of the query. Relatedly, have you had a chance to chat with @geropl about the query's performance, and how to efficiently index and shard it? 👀

easyCZ replied:
You can only use group-by with summary functions (like count). That's because a group-by takes multiple rows, and turns them into a single row by aggregating them.

In practice, we need the full dataset anyway because the billing controller needs to do another pass to drop WSI which have already been billed. 

Happy to go into more details on this.

easyCZ commented:
I've moved this to Draft as I've added a bunch of fixes for flaky tests which I need to pull into a separate PR. The main logic is still reviewable, but I'll move it back to review once cleaned up.

easyCZ commented:
Fixed up and rebased, ready for review again.

geropl commented:
Starting to review now...

geropl commented on components/usage/pkg/controller/reconciler.go:
```diff
@@ -108,129 +102,38 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
-	}
+	for attribution, instances := range u {
+		entity, id := attribution.Values()
+		if entity != db.AttributionEntity_Team {
```
nit: A comment that we handle this in the future would be golden :+1:

geropl commented on components/usage/pkg/controller/reconciler.go:
```diff
@@ -108,129 +102,38 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
-			TeamID:     membership.TeamID,
-			Workspaces: workspacesByOwnerID[userID],
-		})
+		attributedUsage[id] = int64(runtime)
```
This feels dangerous. Why not make attributedUsage accumulate in `uint64`?

easyCZ replied:
Happy to update to unit64 everyone (but Stripe only accepts int64, not unit64). But the reason for using it here is to limit changes to the existing Usage Report reconcile logic with stripe in this PR. That happens here https://github.com/gitpod-io/gitpod/pull/10938/files#diff-b2499b7086d5733f081dcce586e0ff0e77206dd5d8fd6ede638e3fc8f284c798L25-L32 and it currently has int64 as the interface.
I'll update these in a follow-up PR (minus the Stripe API call which will have to be int64) if that's acceptable.

geropl replied:
Ok, fine with this for now.
